data/aws/s3_shared_functions.py

```diff
@@ -312,36 +493,283 @@ def download_s3_file(s3_client, bucket_name, s3_file_key, target_file_path):
+        # if it finds it, it returns a nice user friendly message
+        return_msg, ___ = awssf.aws_exception_handler(ex)
+        raise Exception(return_msg)
+
+
+# -------------------------------------------------
+# You generally want to use this only when you have a large number of files to upload.
+# Size of each file is not relavent.
+# This uses multi-threading and not multi-proc
+def download_folders(
+    s3_client, bucket_name, s3_src_path, trg_folder_path, list_of_search_key=[""], num_workers=1
+):
     """
     Process:
-        - Note: The boto3.client must be already instantated and passed in
+
+        This can be for one folder or multiple folders via search keys.
+
+        - Note: The boto3.client must be already instantiated and passed in
         - Using the incoming s3 src folder, call get_records to get a list of child folders and files
         - Open a s3 client and iterate through the files to download
         - Is recursive
+        - You can optionaly use a list of search keys (see examples below). If you do use non empty
+          search keys, performance will be a bit slower.
+        - If you want everything in a folder, do not submit the list_of_search_keys and it will
+          give you all files under the s3_src_path. This will generally download faster.
     Inputs:
         - s3_client: my_client = boto3.client(profile, creds, whatever)
         - s3_src_path: e.g. /inputs/fema (from s3://{some_bucket}/inputs/fema)
         - trg_folder_path: e.g . /data/inputs/fema
+        - list_of_search_key: simple list of strings, see search key below
+        - num_of_workers if you want to use MT (multi-threading)
 
     Returns:
-        - True or False (did a least one file download successfully)
+        - Number of files downloaded after searching if applicable
 
         The calling code and decide what to do with it.
         Note: Exceptions can still be thrown for catastropic errors (creds, other)
     """
 
-    # TODO: add aws exception handling
+    # Examples for each item in the list of search_keys:
+    #   */hydrotable.csv  - looks at the first level only (ie. huc level)
+    #   */branches/*/rem_zeroed_masked_*.tif  - looks for all branch level rems
+    #   fim_inputs.csv  - file at the root src_s3_folder_path (ie. hand_4_8_7_2)
+    # if list_of_search_key is None or empty, This will just download an entire directory
 
-    # re-validate the connection and credentials as well
-    is_success, return_msg = does_s3_bucket_exist(s3_client, bucket_name)
-    if not is_success:
-        # In this case, we want to raise a new Exception
+    # Note: untested if you want all all files under a subfolder. Might need to use download_folder, TBD
+
+    # if list_of_search_key is None:
+    #     list_of_search_key = [""]
+
+    if not isinstance(list_of_search_key, list):
+        raise Exception("Error: argument for list_of_search_keys must be a 'list' type")
```

Comment:
In this case you should raise a TypeError since the reason is specifically referring to the variable being the wrong type.